hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/exec/systemcontracts/hts/cancelairdrops/TokenCancelAirdropTranslatorTest.java

```diff
@@ -0,0 +1,301 @@
+                configuration);
+
+        // when:
+        boolean matches = subject.matches(attempt);
+
+        // then:
+        assertFalse(matches);
+    }
+
+    @Test
+    void matchesHRCCancelNFTAirdropEnabled() {
+        // given:
+        given(configuration.getConfigData(ContractsConfig.class)).willReturn(contractsConfig);
+        given(contractsConfig.systemContractCancelAirdropsEnabled()).willReturn(true);
+        given(enhancement.nativeOperations()).willReturn(nativeOperations);
+        attempt = prepareHtsAttemptWithSelectorForRedirectWithConfig(
+                TokenCancelAirdropTranslator.HRC_CANCEL_AIRDROP_NFT,
+                subject,
+                enhancement,
+                addressIdConverter,
+                verificationStrategies,
+                gasCalculator,
+                configuration);
+
+        // when:
+        boolean matches = subject.matches(attempt);
+
+        // then:
+        assertTrue(matches);
+    }
+
+    @Test
+    void matchesHRCCancelNFTAirdropDisabled() {
+        // given:
+        given(configuration.getConfigData(ContractsConfig.class)).willReturn(contractsConfig);
+        given(contractsConfig.systemContractCancelAirdropsEnabled()).willReturn(false);
+        given(enhancement.nativeOperations()).willReturn(nativeOperations);
+        attempt = prepareHtsAttemptWithSelectorForRedirectWithConfig(
+                TokenCancelAirdropTranslator.HRC_CANCEL_AIRDROP_NFT,
+                subject,
+                enhancement,
+                addressIdConverter,
+                verificationStrategies,
+                gasCalculator,
+                configuration);
+        // when:
+        boolean matches = subject.matches(attempt);
+
+        // then:
+        assertFalse(matches);
+    }
+
+    @Test
+    void gasRequirementCalculatesCorrectly() {
+        long expectedGas = 1000L;
+        given(gasCalculator.gasRequirement(transactionBody, DispatchType.TOKEN_CANCEL_AIRDROP, payerId))
+                .willReturn(expectedGas);
+
+        long result = TokenCancelAirdropTranslator.gasRequirement(transactionBody, gasCalculator, enhancement, payerId);
+
+        assertEquals(expectedGas, result);
+    }
+
+    @Test
+    void callFromHtsCancelAirdrop() {
+        // given:
+        given(addressIdConverter.convertSender(any())).willReturn(SENDER_ID);
+        given(verificationStrategies.activatingOnlyContractKeysFor(any(), anyBoolean(), any()))
+                .willReturn(verificationStrategy);
+        attempt = prepareHtsAttemptWithSelectorAndCustomConfig(
+                TokenCancelAirdropTranslator.CANCEL_AIRDROP,
+                subject,
+                enhancement,
+                addressIdConverter,
+                verificationStrategies,
+                gasCalculator,
+                configuration);
+
+        // when:
+        var call = subject.callFrom(attempt);
+
+        // then:
+        assertEquals(DispatchForResponseCodeHtsCall.class, call.getClass());
```

Comment:
Only verification is that you get back the right class ... seems insufficient given `TokenCancelAirdropDecoder` L61-71 has that choice of calling three different decoder methods which return three different (though similar!) things.  _That's_ what this unit test should be testing: That the right decoder was called.

In fact since the decoder used here is a mock I'm not even sure _what's_ being called.  I mean, `callFrom` is calling into the decoder: what's happening when those calls are made to the mock?  I thought mockito complained when you called a method you didn't supply a given for (to just call the real method I thought you must use `thenCallRealMethod()` - but maybe not.

Anyway, maybe these tests can use `verify` to check that at least the right decoder was called?

Reply:
Added verify checks